deploy_tools/deploy.py

```diff
@@ -1,14 +1,69 @@
+from typing import Dict
+from web3.contract import Contract
 from web3 import Web3
+from web3.utils.transactions import fill_nonce
 
 
-def deploy_compiled_contract(*, abi, bytecode, web3, constructor_args=()):
+def deploy_compiled_contract(
+    *,
+    abi,
+    bytecode,
+    web3: Web3,
+    constructor_args=(),
+    transaction_options: Dict = None,
+    private_key=None,
+) -> Contract:
+    """
+    Deploys a compiled contract either using an account of the node, or a local private key
+
+    Returns: The deployed contract as a web3 contract
+
+    """
     contract = web3.eth.contract(abi=abi, bytecode=bytecode)
-    txhash = contract.constructor(*constructor_args).transact()
-    receipt = wait_for_successful_transaction_receipt(web3, txhash)
+    constuctor_call = contract.constructor(*constructor_args)
+
+    receipt = send_contract_call(
+        constuctor_call,
+        web3=web3,
+        transaction_options=transaction_options,
+        private_key=private_key,
+    )
+
     address = receipt["contractAddress"]
     return contract(address)
 
 
+def send_contract_call(
+    contract_call, *, web3: Web3, transaction_options: Dict = None, private_key=None
+):
+    """
+    Sends a contract call (for example created with `contract.functions.foo()` as a transaction,
+    either using an account of the node, or a local private key to sign the transaction.
+
+    Returns: The transaction receipt
+
+    """
+    if transaction_options is None:
+        transaction_options = {}
+
+    if private_key is not None:
+        account = web3.eth.account.privateKeyToAccount(private_key)
+        if "from" in transaction_options:
+            raise ValueError(
```

Comment:
Why not rather check if `from == account.address`

Reply:
from is usually not set. I can change it if you want to allow that people can set it to account.address (even though this is useless):
`if "from" in options and options['from'] != account.address`?